pkg/flow/internal/controller/component.go

```diff
@@ -148,6 +150,9 @@ func getManagedOptions(globals ComponentGlobals, cn *ComponentNode) component.Op
 		Logger:        log.With(globals.Logger, "component", cn.nodeID),
 		DataPath:      filepath.Join(globals.DataPath, cn.nodeID),
 		OnStateChange: cn.setExports,
+		Registerer: prometheus.WrapRegistererWith(prometheus.Labels{
+			"component_id": cn.nodeID,
+		}, globals.Registerer),
```

Comment:
Generally I feel like passing the registerer along to components is going to be a common source of bugs, where metrics start leaking because the component implementation didn't unregister them consistently. 

I would much rather us pass a custom implementation of `Registerer` which implements Collector and is invoked by the controller when collecting metrics instead of actually registering metrics globally, so that way components don't need to unregister anything.

Reply:
Added a custom registerer that handles unregistering the component and any children objects.

Reply:
Added a custom wrapper that feeds to the collector and feeds to the prometheus registry.